design/authorization.md

```diff
@@ -0,0 +1,85 @@
+﻿Authorization
+===========
+This document aims at providing some ideas regarding Nessie and authorization, namely:
+* what Nessie authorization should generally cover
+* a Java interface to perform access control, to be used by the JAX-RS server implementation
+* a proposal for an implementation to be bundled with Nessie service
+
+
+## Authorization scope
+It is important to remember that Nessie does not store data directly but only data location and other metadata. As a consequence Nessie authorization layer can only really control access to metadata, but might not prevent data itself to be accessed directly without interacting with Nessie. It is then expected that another system can control access to data itself to make sure no authorized access is made.
+The same is true for access to historical data which is one of Nessie main features. For example while it might seem safe when detecting some undesired sensitive data to commit a change to remove it and to restrict access to only the latest version of the dataset, the truth is that the sensitive data may still exist on the datalake and be accessed by other means (similar to how redacting a PDF by adding black boxes on top of sensitive information does not prevent people to read what is written beneath in most cases). The only safe method to remove this data is to remove it from the datalake, and to add commits to Nessie to point to the new data location.
+
+## Stories
+Here's a list of common authorization scenarios:
+
+* Alice attempts to execute a query against the table `Foo` on branch `prod`. As she has read access to the table on this branch, Nessie allows the execution engine to get the table details so it can locate the data and try to access it.
+* Bob attempts to execute a query against the table `Foo` on branch `prod`. However Bob does not have read access to the table. Nessie returns an authorization error, and the execution engine refuses to execute the query.
+* Carol has access to the branch `prod` content, but not to the table `Foo` on this branch. Carol creates a new reference named `carol-branch`with the same hash as `prod`, and attempts to change permission on table `Foo`. However request is denied, and Carol cannot access the content of `Foo`.
+* Dave has access to the branch `prod` content, and wants to update the content of the table `Foo`. He creates a new reference named `dave-experiment`, and executes several queries against this branch to modify table `Foo`. Each modification is a commit done against `dave-experiment` branch which is approved by Nessie server. When all the desired modifications are done, Dave attempts to merge the changes back to the `prod` branch. However Dave doesn't have the rights to modify the `prod` branch, causing Nessie to deny the request.
```

Comment:
What happens if Dave changes the perms on `dave-experiment` to be more restrictive, and then merges back to `prod`? Which set of perms would then "win"?